### ISSUE TITLE
Is there a mechanism to make notmuch's virtual-folders be synced with the actual maildirs?

### ISSUE THREAD

**Issue by doronbehar** [neomutt/neomutt#1026]
This is more of a question than a bug report / feature request but perhaps it should become a feature request so bare with me.

I use [`mbsync`](http://isync.sourceforge.net) to sync my `maildir` and I use `neomutt`'s together with `notmuch`'s virtual-mailboxes to index my mails etc.
Sometimes, I modify my `+INBOX` (defined with the regular `mailboxes` command) and I want to view afterwards a `virtual-mailbox` which it's `notmuch`'s query returns results which are based on it's current database. It means that if I edit a `mailbox` and than open a `virtual-mailbox`, `notmuch`'s database is not updated and the messages I see are not there. The problem I'm trying to explain is that there is no way (and I've searched throughly the documentation for it) to update `notmuch`'s database with `notmuch new` every time a `mailbox` is changed. Am I missing something?

**Comment by westernmagic** [comment]
The following workarounds come to mind:
 - Watch your maildir with **[`inotify-tools`](https://github.com/rvoicilas/inotify-tools/wiki#info)**
 - Modify your keybindings to run `notmuch new`

**Comment by doronbehar** [comment]
Modifying the key bindings was a good idea, it works great but I still
think (neo)mutt should support some kind of hook mechanism for when the
user is changing the current folder viewed - perhaps a global hook named
`sync-mailbox-hook`.

**Comment by gzagatti** [comment]
I have the same problem. And I agree with @doronbehar, it would be great if neomutt could support some kind of mechanism such as hooks for after it has modified a given mailbox directory, such that other tools know when to re-sync.

Is `inotify-tools` still the best option in this case?

**Comment by gzagatti** [comment]
Similarly, when I am in a virtual folder and try to modify a label with `edit-label`, the label is modified only in the virtual folder. I can even look at the raw message and the label is indeed modified. However, even after syncing the mailbox, looking in the original file or running offlinemap `X-Label` is not modified.

**Comment by flatcap** [comment]
There are quite a few NeoMutt developers who use Notmuch, but they all seem too busy to contribute code (or even ideas).
I'm open to any suggestions, but I don't have the time to implement them.

Feel free to use this issue to brainstorm ideas.
I'm not a Notmuch user, so I don't really know _what_ you need.

**Comment by gzagatti** [comment]
Thanks, I'd like to start contributing but I'm not confident about neomutt's codebase yet. 

Once I have a working configuration and start getting more comfortable with the software I will look into contributing with ideas and code.

**Comment by flatcap** [comment]
> I'd like to start contributing but I'm not confident about neomutt's codebase yet.

There's a lot of background info in our [devel section](https://neomutt.org/dev.html) and all the [functions are now documented](https://neomutt.org/code/).

We also have an IRC channel if you want to ask questions: **#neomutt** on **irc.freenode.net** (due to spamming, you'll need to register)

**Comment by Austin-Ray** [comment]
@doronbehar @gzagatti Would a `folder-hook` work for your use cases?

I was able to craft a `folder-hook`that would call my tagging script whenever I opened a `virtual-mailboxes`. I've been able to bi-directionally sync read status between `maildir` and `virtual-mailboxes` as well as tag any mails that arrived while I was in my `maildir`. 

```rc
# Re-tag on folder open
folder-hook notmuch:// 'exec "/home/aray/.neomutt/notmuch-tag.sh > /dev/null"'
```

This hook will regex match all `virtual-mailboxes` since `notmuch://` is required in the declaration.

**Comment by doronbehar** [comment]
Seems like a good start to me! Does it run when the folder that matches the first argument is being closed or opened? Specifically for our issue of integration lack between `notmuch` and Neomutt this could work good enough just for opening these folders, but it could be great if it could be specified as well the action on which the hook should run.

**Comment by Austin-Ray** [comment]
> The 'Mailbox Matching in Hooks' section doesn't mention support for notmuch://..

An unfortunate lapse in documentation. I only discovered the functionality last night after realizing the examples are `mailboxes` paths! I just have to figure out how to edit the website now :)

> Perhaps... it finished too fast?

The `folder-hook` I wrote pipes `stdout` to `/dev/null` so take that into account.

> does it mean it runs completely in the background

I'm afraid I cannot say. My hypothesis is that push only operates on NeoMutt functions, which is why we need `<shell-escape>` and that brings up a visible shell. However, `exec` appears to be able to call shell functions directly and doesn't display the shell unless you include `shell-escape`. Though `push` and `exec` should be functionally equivalent[0].

[0] https://neomutt.org/guide/configuration.html#21-%C2%A0executing-functions

**Comment by doronbehar** [comment]
> folder-hook I wrote pipes stdout to /dev/null so take that into account.

I didn't pipe my command to `/dev/null` and still couldn't see any shell messages.

> > does it mean it runs completely in the background

> I'm afraid I cannot say.

This should be documented as well..

Thanks for your support!

**Comment by flatcap** [comment]
> My hypothesis is that push only operates on NeoMutt functions,

The `push` command puts stuff directly into the keyboard buffer.
`<...>` sequences are interpreted as commands, but the rest is literal. 

> I just have to figure out how to edit the website now :)

The source of the guide is `doc/manual.xml.head` in the code repo.
It's currently in DocBook (xml) format.

How that gets to the website is complicated: see the right-hand diagram, [here](https://neomutt.org/dev/make).
Currently it involves a load of manual work (for me) in vim.

**Comment by Austin-Ray** [comment]
> The push command puts stuff directly into the keyboard buffer.

Okay! I understand. It's essentially scripting keyboard presses with support for key-binding agnostic function invocation. `push '!echo "hello,world'"` == `push '<shell-escape>echo "hello, world"'`

Despite the documentation implies about `exec` and `push` being equivalent[0], there are behavioral differences that need be documented. I will make a note to clarify it.

Behavior differences:
1. `exec` does not output to `stdout`
2. `exec`, when issued through the UI, is limited to NeoMutt functions.
3. Different behavior depending on if its in the configuration (my `folder-hook`'s `exec` doesn't work if I type it in due to the above point)

Are any of these bugs, @flatcap? The third point seems like one.

I apologise for getting a bit off-topic. 

[0] https://neomutt.org/guide/configuration.html#21-%C2%A0executing-functions

**Comment by flatcap** [comment]
> documentation implies about `exec` and `push` being equivalent

Hmm... the docs are a bit misleading.
`exec` is limited to functions, so `push` emulate `exec` but not the other way around.

> `exec` does not output to `stdout`

Hmm... I hadn't noticed.  I wonder if that's by design.
(or if the messages are getting cleared before the user sees them -- check the logs)

With `push`, NeoMutt doesn't see a difference between _that_ and the user pressing the keys.

> my `folder-hook`'s `exec` doesn't work

I'm not sure anyone understands the abilities and limitations of hooks.
Put your broken hook in an issue and we'll start investigating (and documenting).

> I apologise for getting a bit off-topic.

np

**Comment by Austin-Ray** [comment]
@doronbehar I apologise for the off-topic conversation in your issue. Will using a `folder-hook` be a sufficient work-around until a proper mechanism is implemented?

---

@flatcap 
> ... your broken hook

The hook itself works and executes with no problem. However, if I try to `exec "/home/aray/.neomutt/notmuch-tag.sh"` through the UI, I receive `"/home/aray/.neomutt/notmuch-tag.sh: no such function"`. When declared in the configuration, `exec` can execute any arbitrary function/shell script, but through the UI, it is limited NeoMutt's built-in functions. 

> Hmm... I hadn't noticed. I wonder if that's by design.
> (or if the messages are getting cleared before the user sees them -- check the logs)

I haven't checked the log, but I am unable to redirect output to a specific file, which makes me suspect that `exec` doesn't have output at all. I'll continue looking.

> .... an issue and we'll start investigating (and documenting).

Will do.

**Comment by doronbehar** [comment]
No need to apologize @Austin-Ray, if it hadn't for you we wouldn't have discussed `folder-hook` at all, I've just started having fun :)

As for our original issue, `folder-hook` could certainly be a sufficient work-around and this is what I was looking for originally when I opened this issue since the documentation wasn't clear enough for me.
But, Now that I'm thinking about it, I must say that IMO such a mechanism shouldn't be a new configuration option but a built in behaviour people should accept and get used to - the `notmuch` back-end of Neomutt should automatically sync `notmuch`'s index with the mailbox after every operation performed in it, this shouldn't be done with shell commands or shell scripts but right through `libnotmuch.so`. Very much like what the `notmuch new` command does.

To summarize:

- The behaviour of `exec` and `folder-hook` should be further investigated and better documented.
- An additional behaviour to Neomutt's `notmuch` back-end should be added that will sync every operation done to any mailbox with `notmuch`'s index.

BTW:

This is the [2nd issue I've opened here](https://github.com/neomutt/neomutt/issues/1014) which is to be solved through documentation improvement. https://github.com/neomutt/neomutt.github.io/pull/73 Fixed it then but I'm beginning to feel like the documentation is very uninviting for contributes. Have you ever considered switching to https://readthedocs.org/? It has many users and many people are familiar with it's documents' format which is definitely much more flexible and modular.

**Comment by doronbehar** [comment]
> But... we still need an easily-browsed format for the documentation installed with NeoMutt.

I totally agree with you. I was thinking about this as well in regard to other projects that use readthedocs.org and since it is based on [Sphinx](http://www.sphinx-doc.org/), it is possible to create `info` documents which can be very useful for a command line program. See https://github.com/rtfd/readthedocs.org/issues/4458, especially  [the currently last comment](https://github.com/rtfd/readthedocs.org/issues/4458#issuecomment-427671474).

**Comment by flatcap** [comment]
Hmm... I hadn't considered `info` (though personally, I don't like it -- I can never figure out navigation).

Here's my plan, in brief:

### Requirements

- One source for all docs
- Markdown format
- Easy user contribution

### Plan

- Separate doc repo
  - Markdown files (with hyperlinks)
  - Nested dirs (for sections)
  - YAML header for metadata
  - Accept Issues and Pull Requests

- Website Guide
  - Use GitHub pages (Jeykll)
  - Each page has a button: "Improve this page" (which links to GitHub editor)
  - Use Algolia to search

- Installed Guide
  - Install guide repo
  - Add a new "Help" backend to NeoMutt
  - Use NeoMutt to view/search the guide
  - Use `links` navigation style (arrow keys to select and follow hyperlinks)
  - Each page has a button "Improve this page" (which opens the compose window)

### Benefits

- The user would see a consistent guide
- The user would only ever be one click away from contributing
- Easy maintenance
- No complex deployment steps

**Comment by thomasjfox** [comment]
> Here's my plan, in brief:
> ### Requirements
>     * Markdown format

one thing regarding the markdown format: How should future translations of the manual be handled?
How does a translator get updated strings / paragraphs without looking at git diffs?

Last year I've checked out multiple tools to extract strings for gettext directly from docbook documents. The final tool used to translate a 450+ page technical manual was "xml2po" from gnome-doc-utils. It even inserts dummy strings with md5sums of pictures. So when a screenshot changes, the translator gets the "translation" marked as fuzzy and can take a look at the updated source screenshot.

tl;dr: Is there a tool to easily translate markdown content?

**Comment by flatcap** [comment]
> > markdown

DocBook just isn't a viable option for a crowd-sourced documentation system.
Behind the scenes it's currently a ****ing nightmare of scripts and manual steps to turn it into valid, correctly-linked, HTML.

> translation

I hadn't thought very far ahead for translation of the docs.
Twenty years Mutt's been around and nobody's attempted to translate the main guide.
By splitting it into tiny Markdown snippets, I might be able to encourage someone to try.

> ... easily translate ...

No.  Nothing ever is :-)

> How does a translator get updated strings / paragraphs without looking at git diffs?

It'll always be diffs in some form or other.

> tool used to translate ... was "xml2po" from gnome-doc-utils

Sounds amazing.  But without someone willing to dedicate a large amount of their time to it, it's not going to happen.

**Comment by doronbehar** [comment]
Perhaps not being able to translate the documentation is a price we should be willing to pay at least for the start. After all, Neomutt is a power users' email client. IMO it'd be rather dissent to expect our users to know English good enough at least for reading the documentation. If they hadn't known English at all they'd have used Evolution :wink:.
